File: CK3toEU4/Source/CK3World/Religions/Religions.cpp
#include "Religions.h"
#include "Religion.h"
#include "Log.h"
#include "ParserHelpers.h"

CK3::Religions::Religions(std::istream& theStream)
{
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}

void CK3::Religions::registerKeys()
{
	registerRegex(R"(\d+)", [this](const std::string& faithID, std::istream& theStream) {
		auto newReligion = std::make_shared<Religion>(theStream, std::stoll(faithID));
		religions.insert(std::pair(newReligion->getID(), newReligion));
	});
	registerKeyword("religions", [this](const std::string& unused, std::istream& theStream) {
		religions = Religions(theStream).getReligions();
	});
	registerKeyword("faiths", [this](const std::string& unused, std::istream& theStream) {
		faiths = Faiths(theStream);
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

void CK3::Religions::linkFaiths(const Faiths& theFaiths)
{
	auto counter = 0;
	const auto& faithData = theFaiths.getFaiths();
	for (const auto& religion: religions)
	{
		const auto& religionFaiths = religion.second->getFaiths();
		std::map<long long, std::shared_ptr<Faith>> replacementMap;
		
		for (const auto& faith: religionFaiths)
		{
			const auto& faithDataItr = faithData.find(faith.first);
			if (faithDataItr != faithData.end())
			{
				replacementMap.insert(*faithDataItr);
			}
			else
			{
				throw std::runtime_error(
					 "Religion " + religion.second->getName() + " has faith " + std::to_string(faith.first) + " which has no definition!");
			}			
		}
		religion.second->loadFaiths(replacementMap);
		++counter;
	}
	Log(LogLevel::Info) << "<> " << counter << " religions updated.";
}

File: CK3ToEU4/Source/CK3/Religions/Religions.cs
using System;
using System.Collections.Generic;
using commonItems;

namespace CK3ToEU4.CK3.Religions;

class Religions
{
	public Religions(BufferedReader reader)
	{
		var parser = new Parser();
		registerKeys(parser);
		parser.ParseStream(reader);
	}

	public IReadOnlyDictionary<long, Religion> getReligions() { return religions; }
	public Faith getFaiths() { return faiths; } // Use this only once in World.cpp

	void linkFaiths(Faiths theFaiths)
	{
		int counter = 0;
		var faithData = theFaiths.getFaiths();
		foreach (var religion in religions)
		{
			var religionFaiths = religion.Value.getFaiths();
			Dictionary<long, Faith> replacementMap;

			foreach (var faith in religionFaiths)
			{
				const auto& faithDataItr = faithData.find(faith.first);
				if (faithDataItr != faithData.end())
				{
					replacementMap.insert(*faithDataItr);
				}
				else
				{
					throw new Exception($"Religion {religion.Value.getName()} has faith {faith.Key} which has no definition!");
				}
			}
			religion.Value.loadFaiths(replacementMap);
			++counter;
		}
		Logger.Info("<> " + counter + " religions updated.");
	}

	private void registerKeys(Parser parser)
	{
		parser.RegisterRegex(CommonRegexes.Integer, (reader, faithID) => {
			var newReligion = new Religion(reader, long.Parse(faithID));
			religions.Add(newReligion.ID, newReligion);
		});
		parser.RegisterKeyword("religions", reader => {
			religions = new(new Religions(reader).getReligions());
		});
		parser.RegisterKeyword("faiths", reader => {
			faiths = new Faiths(reader);
		});
		parser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreItem);
	}

	Dictionary<long, Religion> religions;
	Faiths faiths;
};